can/interfaces/multicast_ip/bus.py

```diff
@@ -0,0 +1,338 @@
+        """
+        self._multicast.shutdown()
+
+    @staticmethod
+    def _detect_available_configs() -> List[AutoDetectedConfig]:
+        if hasattr(socket, "CMSG_SPACE"):
+            return [
+                {
+                    "interface": "multicast_ip",
+                    "channel": MulticastIpBus.DEFAULT_GROUP_IPv6,
+                },
+                {
+                    "interface": "multicast_ip",
+                    "channel": MulticastIpBus.DEFAULT_GROUP_IPv4,
+                },
+            ]
+
+        # else, this interface cannot be used
+        return []
+
+
+class GeneralPurposeMulticastIpBus:
+    """A general purpose send and receive handler for multicast over IP/UDP."""
+
+    def __init__(
+        self, group: str, port: int, hop_limit: int, max_buffer: int = 4096
+    ) -> None:
+        self.group = group
+        self.port = port
+        self.hop_limit = hop_limit
+        self.max_buffer = max_buffer
+
+        # Look up multicast group address in name server and find out IP version of the first suitable target
+        # and then get the address family of it (socket.AF_INET or socket.AF_INET6)
+        connection_candidates = socket.getaddrinfo(
+            group, self.port, type=socket.SOCK_DGRAM
+        )
+        sock = None
+        for connection_candidate in connection_candidates:
+            address_family: socket.AddressFamily = connection_candidate[0]
+            self.ip_version = 4 if address_family == socket.AF_INET else 6
+            try:
+                sock = self._create_socket(address_family)
+            except OSError as error:
+                log.info(
+                    f"could not connect to the multicast IP network of candidate %s; reason: {error}",
+                    connection_candidates,
+                )
+        if sock is not None:
+            self._socket = sock
+        else:
+            raise RuntimeError("could not connect to a multicast IP network")
+
+        # used in recv()
+        self.received_timestamp_struct = "@II"
+        ancillary_data_size = struct.calcsize(self.received_timestamp_struct)
+        self.received_ancillary_buffer_size = socket.CMSG_SPACE(ancillary_data_size)
+
+        # used by send()
+        self._send_destination = (self.group, self.port)
+        self._last_send_timeout: Optional[float] = None
+
+    def _create_socket(self, address_family: socket.AddressFamily) -> socket.socket:
+        """Creates a new socket. This might fail and raise an exception!
+
+        :param address_family: whether this is of type `socket.AF_INET` or `socket.AF_INET6`
+        :raises OSError: if the socket could not be opened or configured correctly; in this case, it is
+                         guaranteed to be closed/cleaned up
+        """
+        # create the UDP socket
+        # this might already fail but then there is nothing to clean up
+        sock = socket.socket(address_family, socket.SOCK_DGRAM)
```

Comment:
Using UDP causes fire and forget. Possible issues here:
- How to know when a CAN message is transmitted on the "bus" successfully?
- How to know when a CAN message is lost? In hardware CAN controllers this is not possible, a message is either transmitted on the bus, or in the queue.

Reply:
You are right, this is potentially problematic. On the other hand, Multicast + normal TCP seems to be a weird couple too, as basic multicast is inherently unreliable. We don't even know the recipients of our messages. I'd leave it as is, and maybe put a bigger disclaimer in the docs? In practice and in many local/small networks, UDP it's definitely reliable enough for this. And on a loopback/link-local network too. And that's one major area of application.

Reply:
Agreed to put a disclaimer for this in the docs. Also, I'm not sure how often this would happen, and if so, if it poses an issue. Maybe it does pose and issue, since the CAN bus is expected to be reliable, so sending a CAN message does not deal with re-transmits and all this stuff. You either send a message, or you do not. This is a good feature of the CAN protocol.